Documentation/ApiOverview/Localization/LocalizationApi/LanguageService.rst

```diff
@@ -12,5 +12,37 @@ be created directly, therefore its constructor is internal. Create a
 
 In the backend context a :php:`LanguageService` is stored in the global
 variable :php:`$GLOBALS['LANG']`.
+In the frontend a :php:`LanguageService` can be accessed via the contentObject:
+
+..  code-block:: php
+    :caption: Classes/Controller/ExampleController.php
+
+    use Psr\Http\Message\ServerRequestInterface;
+    use TYPO3\CMS\Core\Localization\LanguageServiceFactory;
+    use TYPO3\CMS\Core\Utility\GeneralUtility;
+
+    class ExampleController {
+        proteced ServerRequestInterface $request;
+
+        public function processAction(ServerRequestInterface $request): string
+            $this->request = $request;
+
+            $content = '';
+            ...
+            $label = $this->getTranslatedLabel('LLL:EXT:my_extension/Resources/Private/Language/locallang.xlf:labels.exampleLabel');
+            ...
+
+            return $content;
+        }
+
+        protected function getTranslatedLabel(string $key): string
```

Comment:
```suggestion
        private function getTranslatedLabel(string $key): string
```

Reply:
Cool idea to make XCLASS system in TYPO3 useless ;-)

Reply:
Irony can be hard to interpret, please let us try to be unoppionated documentors here :-D 

The "final" and "private" opinions can be very splitting, both sides have understandable arguments. I think by XCLASSing or extending extensions, the maintainer needs to specifically "opt in" to allowing their code to be extended, otherwise other people start to rely on things that the extension being extended does not want to provide, and breakage occurs, and people get mad.

I don't know though how the rest of the TYPO3 docs handles example code like this; we should use the variant that most of our other recent code examples use, and not make exemptions here at this single place, and probably follow the TYPO3-lead here, were the current practice seems to be to make "helper methods" only be subject to the single controller that it's used in? Otherwise a Service would need to be created, blowing up the scope of the example.

Reply:
Okay, let's differentiate:

1. For a project-specific extension: No need to XCLASS something. 

2. For a third-party extension someone published (when I am the maintainer): If someone wants to extends one of my extension, they contact me. Then we find a solution for that use case (mostly a new PSR-14 event). So no XCLASSing necessary. XCLASSing may also break functionality for the user, as I don't care refactoring my third-party extension also for bugfix releases. So: This is a protection for the integrator.

Lina and I agreed to make classes in examples mostly final (and nowadays readonly when applicable) as best practise. What the user then does when copying/adjusting their code is their thing. And for final classes protected methods are superfluous. I suggested the private visibility because of the final class suggestion.

Reply:
If you write a hook, make it private/final, as you still can override the result with another hook. 

If you write an EventListener, make it private/final, as you still can override the result with another EventListener. 

If you write a Middleware, make it private/final, as you still can override the result with another Middleware. 

If you write a ViewHelper, make it private/final, as you still can write and use your own ViewHelper.

In all the other cases, keep `protected` to give XCLASS and EXT:extender a chance.

Just my two cents. But as I have seen, the decision seems to be done: `Lina and I agreed to make classes in examples mostly final`. I'm with you to declare classes in constructor arguments as `readonly` where ever possible.

Reply:
Yes, Chris and I agreed a while ago to make these settings as strict as possible as a best practice (private, final, readonly wherever applicable) unless there is a good reason that something should be extended.  

I am not sure we ever wrote this descision down, but maybe we should. 

If you want to adjust this descision we can have another discussion round @froemken.

Reply:
Then we should provide a chapter to document what is recommended in which case (final/readonly and public/protected/private), so developers can refer to and rely on it. This should then also be agreed by Core developers IMHO.

Speaking for me: When a user wants to XCLASS a class of my extensions, I don't really care if their XCLASSed code breaks on updates, as I do not consider XCLASSing as API in my extensions. To stress this, I am using "final" classes. If someone wants to change my code on own risk, they can use composer-patches. Or just talk to me and there will be a **future-proof** and **supported** solution (this happened already a few times). Just my 2 cents.